internal/forge/gitlab/comment.go

```diff
@@ -0,0 +1,155 @@
+package gitlab
+
+import (
+	"context"
+	"fmt"
+	"github.com/xanzy/go-gitlab"
+	"go.abhg.dev/gs/internal/forge"
+	"iter"
+	"strconv"
+)
+
+type MRComment struct {
+	Number   int `json:"number"`
+	MRNumber int `json:"mr_number"`
+}
+
+var _ forge.ChangeCommentID = (*MRComment)(nil)
+
+func mustMRComment(id forge.ChangeCommentID) *MRComment {
+	if id == nil {
+		return nil
+	}
+
+	mrc, ok := id.(*MRComment)
+	if !ok {
+		panic(fmt.Sprintf("unexpected MR comment type: %T", id))
+	}
+	return mrc
+}
+
+func (c *MRComment) String() string {
+	return strconv.Itoa(c.Number) // TODO: return URL?
+}
+
+// PostChangeComment posts a new comment on an MR.
+func (r *Repository) PostChangeComment(
+	_ context.Context,
+	id forge.ChangeID,
+	markdown string,
+) (forge.ChangeCommentID, error) {
+	noteOptions := gitlab.CreateMergeRequestNoteOptions{
+		Body: &markdown,
+	}
+
+	mrNumber := mustMR(id).Number
+	note, _, err := r.client.Notes.CreateMergeRequestNote(r.repoID, mrNumber, &noteOptions)
+	if err != nil {
+		return nil, fmt.Errorf("post comment: %w", err)
+	}
+
+	r.log.Debug("Posted comment", "id", note.ID) // TODO URL?
+	return &MRComment{
+		Number:   note.ID,
+		MRNumber: mrNumber,
+	}, nil
+}
+
+func (r *Repository) UpdateChangeComment(
+	_ context.Context,
+	id forge.ChangeCommentID,
+	markdown string,
+) error {
+	mrComment := mustMRComment(id)
+	noteOptions := gitlab.UpdateMergeRequestNoteOptions{
+		Body: &markdown,
+	}
+	_, _, err := r.client.Notes.UpdateMergeRequestNote(r.repoID, mrComment.MRNumber, mrComment.Number, &noteOptions)
+	if err != nil {
+		return fmt.Errorf("update comment: %w", err)
+	}
+
+	r.log.Debug("Updated comment", "id", mrComment.Number) // TODO URL?
+
+	return nil
+}
+
+// There isn't a way to filter comments by contents server-side,
+// so we'll be doing that client-side.
+// GitLab's API paginates the notes listing endpoints so we will fetch them by 20 per page.
+//
+// Since our comment will usually be among the first few comments,
+// that, plus the ascending order of comments, should make this good enough.
+var _listChangeCommentsPageSize = 20 // var for testing
+
+// ListChangeComments lists comments on an MR,
+// optionally applying the given filtering options.
+func (r *Repository) ListChangeComments(
+	ctx context.Context,
+	id forge.ChangeID,
+	options *forge.ListChangeCommentsOptions,
+) iter.Seq2[*forge.ListChangeCommentItem, error] {
+	var filters []func(gitlab.Note) (keep bool)
+	if options != nil {
+		if len(options.BodyMatchesAll) != 0 {
+			for _, re := range options.BodyMatchesAll {
+				filters = append(filters, func(node gitlab.Note) bool {
+					return re.MatchString(node.Body)
+				})
+			}
+		}
+
+		// Only the author or maintainers can update comments.
+		if options.CanUpdate {
+			filters = append(filters, func(note gitlab.Note) bool {
+				return note.Author.ID == r.userID || *r.userRole >= gitlab.MaintainerPermissions
+			})
+		}
```

Comment:
The GitLab API does not provide information about whether you can update a comment or not, so I am calculating it based on this [Edit a comment](https://docs.gitlab.com/ee/user/discussions/#edit-a-comment).
> You can edit your own comment at any time. Anyone with at least the Maintainer role can also edit a comment made by someone else.